### ISSUE TITLE
Unable to load enet image segmentation model on opencv

### ISSUE THREAD

**Issue by AmanGoyalNayan** [opencv/opencv#17128]
##### System information (version)

- OpenCV =>  4.2
- Operating System / Platform => Ubuntu 18.04
- Compiler => Python3.6.9 : Vscode

##### Detailed description

I had worked on enet image segmentation model . Its architecture is given as below :
...python3.6.9
def initial_block(inp):
    inp1 = inp
    conv = Conv2D(filters=13, kernel_size=3, strides=2, padding='same', use_bias=False, kernel_initializer='he_normal')(inp)
    pool = MaxPool2D(2)(inp1)
    concat = concatenate([conv, pool])
    return concat
def encoder_bottleneck(inp, filters, dilation_rate=2, downsample=False, dilated=False, asymmetric=False, drop_rate=0.1):
    reduce = filters // 4
    down = inp
    kernel_stride = 1
    #Downsample
    if downsample:
        kernel_stride = 2
        pad_activations = filters - inp.shape.as_list()[-1]
        down = MaxPool2D(2)(down)
        down = Permute(dims=(1, 3, 2))(down)
        down = ZeroPadding2D(padding=((0, 0), (0, pad_activations)))(down)
        down = Permute(dims=(1, 3, 2))(down)
    #1*1 Reduce
    x = Conv2D(filters=reduce, kernel_size=kernel_stride, strides=kernel_stride, padding='same', use_bias=False, kernel_initializer='he_normal')(inp)
    x = BatchNormalization(momentum=0.1)(x)
    x = PReLU(shared_axes=[1, 2])(x)
    #Conv
    if not dilated and not asymmetric:
        x = Conv2D(filters=reduce, kernel_size=kernel_stride, padding='same', kernel_initializer='he_normal')(x)
    elif dilated:
        x = Conv2D(filters=reduce, kernel_size=kernel_stride, padding='same', dilation_rate=dilation_rate, kernel_initializer='he_normal')(x)
    elif asymmetric:
        x = Conv2D(filters=reduce, kernel_size=(1,5), padding='same', use_bias=False, kernel_initializer='he_normal')(x)
        x = Conv2D(filters=reduce, kernel_size=(5,1), padding='same', kernel_initializer='he_normal')(x)
    x = BatchNormalization(momentum=0.1)(x)
    x = PReLU(shared_axes=[1, 2])(x)
    #1*1 Expand
    x = Conv2D(filters=filters, kernel_size=kernel_stride, padding='same', use_bias=False, kernel_initializer='he_normal')(x)
    x = BatchNormalization(momentum=0.1)(x)
    x = SpatialDropout2D(rate=drop_rate)(x)
    concat = Add()([x, down])
    concat = PReLU(shared_axes=[1, 2])(concat)
    return concat    
def decoder_bottleneck(inp, filters, upsample=False):
    reduce = filters // 4
    up = inp
    #Upsample
    if upsample:
        up = Conv2D(filters=filters, kernel_size=1, strides=1, padding='same', use_bias=False, kernel_initializer='he_normal')(up)
        up = UpSampling2D(size=2)(up)
    #1*1 Reduce
    x = Conv2D(filters=reduce, kernel_size=1, strides=1, padding='same', use_bias=False, kernel_initializer='he_normal')(inp)
    x = BatchNormalization(momentum=0.1)(x)
    x = PReLU(shared_axes=[1, 2])(x)
    #Conv
    if not upsample:
        x = Conv2D(filters=reduce, kernel_size=3, strides=1, padding='same', kernel_initializer='he_normal')(x)
    else:
        x = Conv2DTranspose(filters=reduce, kernel_size=3, strides=2, padding='same', kernel_initializer='he_normal')(x)
    x = BatchNormalization(momentum=0.1)(x)
    x = PReLU(shared_axes=[1, 2])(x)
    #1*1 Expand
    x = Conv2D(filters=filters, kernel_size=1, strides=1, padding='same', use_bias=False, kernel_initializer='he_normal')(x)
    x = BatchNormalization(momentum=0.1)(x)
    concat = Add()([x, up])
    concat = PReLU(shared_axes=[1, 2])(concat)
    return concat
def ENet(inp, nclasses=1):
    enc = initial_block(inp)
    enc = BatchNormalization(momentum=0.1)(enc)
    enc = PReLU(shared_axes=[1, 2])(enc)
    #Bottleneck 1.0
    enc = encoder_bottleneck(enc, 64, downsample=True, drop_rate=0.001)
    enc = encoder_bottleneck(enc, 64, drop_rate=0.001)
    enc = encoder_bottleneck(enc, 64, drop_rate=0.001)
    enc = encoder_bottleneck(enc, 64, drop_rate=0.001)
    enc = encoder_bottleneck(enc, 64, drop_rate=0.001)
    #Bottleneck 2.0
    enc = encoder_bottleneck(enc, 128, downsample=True)
    enc = encoder_bottleneck(enc, 128)
    enc = encoder_bottleneck(enc, 128, dilation_rate=2, dilated=True)
    enc = encoder_bottleneck(enc, 128, asymmetric=True)
    enc = encoder_bottleneck(enc, 128, dilation_rate=4, dilated=True)
    enc = encoder_bottleneck(enc, 128)
    enc = encoder_bottleneck(enc, 128, dilation_rate=8, dilated=True)
    enc = encoder_bottleneck(enc, 128, asymmetric=True)
    enc = encoder_bottleneck(enc, 128, dilation_rate=16, dilated=True)
    #Bottleneck 3.0
    enc = encoder_bottleneck(enc, 128)
    enc = encoder_bottleneck(enc, 128, dilation_rate=2, dilated=True)
    enc = encoder_bottleneck(enc, 128, asymmetric=True)
    enc = encoder_bottleneck(enc, 128, dilation_rate=4, dilated=True)
    enc = encoder_bottleneck(enc, 128)
    enc = encoder_bottleneck(enc, 128, dilation_rate=8, dilated=True)
    enc = encoder_bottleneck(enc, 128, asymmetric=True)
    enc = encoder_bottleneck(enc, 128, dilation_rate=16, dilated=True)
    #Bottleneck 4.0
    dec = decoder_bottleneck(enc, 64, upsample=True)
    dec = decoder_bottleneck(dec, 64)
    dec = decoder_bottleneck(dec, 64)
    #Bottleneck 5.0
    dec = decoder_bottleneck(dec, 16, upsample=True)
    dec = decoder_bottleneck(dec, 16)
    dec = Conv2DTranspose(filters=nclasses, kernel_size=3, strides=2, padding='same', activation='sigmoid')(dec)
    model = Model(inputs=inp, outputs=dec, name='Enet')
    #model.save('enet.h5')
    return model        
inp = Input(shape=(512, 512, 1))
enet = ENet(inp, 1)    
...
Whenever I load .pb file by opencv , I get following error : 
error: OpenCV(4.2.0) /io/opencv/modules/dnn/src/tensorflow/tf_importer.cpp:1244: error: (-212:Parsing error) Only NHWC <-> NCHW permutations are allowed. in function 'populateNet'

Code snippet by which error could be reproduced : 
...
net = cv2.dnn.readNetFromTensorflow('/home/aman/Downloads/yardsegenet.pb')
...
Link to yardsegenet.pb : 

https://drive.google.com/open?id=1qy2eYbSii4Tmopc7OnrpDVfVbeRt-QdG

Whenever I load both .pb and .pbtxt files , I get the following error : 
error: OpenCV(4.2.0) /io/opencv/modules/dnn/src/tensorflow/tf_importer.cpp:663: error: (-215:Assertion failed) const_layers.insert(std::make_pair(name, li)).second in function 'addConstNodes'

The code snippet used to reproduce this error is as follows : 
...
net = cv2.dnn.readNetFromTensorflow('/home/aman/env/yardsegenet.pb','/home/aman/env/pbtxt/yardsegenet.pbtxt')
...
Link to yardsegenet.pbtxt
https://drive.google.com/open?id=15VB0hx7z8O6ArjT0Pg7bSqiI6aloH6Om

I am providing (512,512,1) image size to my model.

I have been working on this since a week now and repetedly getting this error.

**Comment by dkurt** [comment]
Please describe the issue in such way so any user can reproduce it - add minimal code snippet with OpenCV usage, upload used files (`.pb` in that case), provide input tensor resolution.

**Comment by AmanGoyalNayan** [comment]
> Please describe the issue in such way so any user can reproduce it - add minimal code snippet with OpenCV usage, upload used files (`.pb` in that case), provide input tensor resolution.

Hey. I tried adding .pb file. It says wrong format.

**Comment by dkurt** [comment]
@AmanGoyalNayan , choose files sharing service, Dropbox in example. We have no time for remote debugging. The issue description should contain all the details to reproduce it.

Please consider our guidelines and just respect others time.

**Comment by AmanGoyalNayan** [comment]
> @AmanGoyalNayan , choose files sharing service, Dropbox in example. We have no time for remote debugging. The issue description should contain all the details to reproduce it.
> 
> Please consider our guidelines and just respect others time.

Yes I am sorry @dkurt  . Actually I am posting an issue for the first time. Now I have changed my description as per requirements. Do suggest me if something more is required.

**Comment by dkurt** [comment]
Hi!

Can you please explain what actually the following layers do?
```python
down = Permute(dims=(1, 3, 2))(down)
down = ZeroPadding2D(padding=((0, 0), (0, pad_activations)))(down)
down = Permute(dims=(1, 3, 2))(down)
```
